sw/device/lib/base/mmio.h

```diff
@@ -288,6 +288,30 @@ inline void mmio_region_nonatomic_set_bit32(mmio_region_t base,
   mmio_region_nonatomic_set_mask32(base, offset, 0x1, bit_index);
 }
 
+/**
+ * Copies a block of memory from MMIO to main memory while ensuring that MMIO
+ * accesses are word-aligned.
+ *
+ * @param base the MMIO region to read from.
+ * @param offset the offset to start reading from, in bytes.
+ * @param buf the main memory location to start writing to.
+ * @param count number of bytes to copy.
+ */
+void mmio_region_memcpy_from_mmio32(mmio_region_t base, uint32_t offset,
+                                    void *dst, size_t count);
+
+/**
+ * Copies a block of memory from main memory to MMIO while ensuring that MMIO
+ * accesses are word-aligned.
+ *
+ * @param base the MMIO region to write to.
+ * @param offset the offset to start writing to, in bytes.
+ * @param buf the main memory location to start reading from.
+ * @param count number of bytes to copy.
+ */
+void mmio_region_memcpy_to_mmio32(mmio_region_t base, uint32_t offset,
+                                  void *src, size_t count);
```

Comment:
Nit: `src` can be a `const void*`.

Reply:
I was on the fence about this one because the `const` will be casted away to be able to call the implementation but your comment pushed me over :). Done. This should make the public API clearer if nothing else.